go/chat/sender.go

```diff
@@ -81,57 +81,71 @@ func (s *BlockingSender) addSenderToMessage(msg chat1.MessagePlaintext) (chat1.M
-		return chat1.MessagePlaintext{}, fmt.Errorf("addPrevPointersToMessage expects an empty prev list")
+		return resMsg, fmt.Errorf("addPrevPointersToMessage expects an empty prev list")
 	}
 
-	res, err := s.G().ConvSource.Pull(ctx, conv.GetConvID(), msg.ClientHeader.Sender,
-		chat1.GetThreadReason_PREPARE,
-		&chat1.GetThreadQuery{
-			DisableResolveSupersedes: true,
-		},
-		s.prevPtrPagination)
-	if err != nil {
-		return chat1.MessagePlaintext{}, err
-	}
-
-	if len(res.Messages) == 0 {
-		s.Debug(ctx, "no local messages found for prev pointers")
-	}
+	var thread chat1.ThreadView
 	var prevs []chat1.MessagePreviousPointer
-	newPrevsForRegular, newPrevsForExploding, err := CheckPrevPointersAndGetUnpreved(&res)
-	if err != nil {
-		return chat1.MessagePlaintext{}, err
-	}
+	pagination := &chat1.Pagination{
+		Num: s.prevPtrPagination.Num,
+	}
+	numRetries := 5
+	for i := 0; i < numRetries; i++ {
+		thread, err = s.G().ConvSource.Pull(ctx, conv.GetConvID(), msg.ClientHeader.Sender,
+			chat1.GetThreadReason_PREPARE,
+			&chat1.GetThreadQuery{
+				DisableResolveSupersedes: true,
+			},
+			pagination)
+		if err != nil {
+			return resMsg, err
+		} else if thread.Pagination == nil {
+			break
+		}
+		pagination.Next = thread.Pagination.Next
 
-	var hasPrev bool
-	if msg.IsEphemeral() {
-		prevs = newPrevsForExploding
-		hasPrev = len(newPrevsForExploding) > 0
-	} else {
-		prevs = newPrevsForRegular
-		// If we have only sent ephemeralMessages and are now sending a regular
-		// message, we may have an empty list for newPrevsForRegular. In this
-		// case we allow the `Prev` to be empty, so we don't want to abort in
-		// the check on numPrev below.
-		hasPrev = len(newPrevsForRegular) > 0 || len(newPrevsForExploding) > 0
-	}
+		if len(thread.Messages) == 0 {
+			s.Debug(ctx, "no local messages found for prev pointers")
+		}
+		newPrevsForRegular, newPrevsForExploding, err := CheckPrevPointersAndGetUnpreved(&thread)
+		if err != nil {
+			return resMsg, err
+		}
 
-	if !hasPrev {
-		return chat1.MessagePlaintext{}, fmt.Errorf("Could not find previous messages for prev pointers (of %v)", len(res.Messages))
+		var hasPrev bool
+		if msg.IsEphemeral() {
+			prevs = newPrevsForExploding
+			hasPrev = len(newPrevsForExploding) > 0
+		} else {
+			prevs = newPrevsForRegular
+			// If we have only sent ephemeralMessages and are now sending a regular
+			// message, we may have an empty list for newPrevsForRegular. In this
+			// case we allow the `Prev` to be empty, so we don't want to abort in
+			// the check on numPrev below.
+			hasPrev = len(newPrevsForRegular) > 0 || len(newPrevsForExploding) > 0
+		}
+
+		if hasPrev {
+			break
+		} else if thread.Pagination.Last {
+			return resMsg, fmt.Errorf("Could not find previous messages for prev pointers (of %v)", len(thread.Messages))
+		} else {
+			s.Debug(ctx, "Could not find previous messages for prev pointers (of %v), attempt: %v, retrying", len(thread.Messages), i)
```

Comment:
It looks like the intention is that after hitting this case 5 times, the loop just exits, and allows the prev list to be empty. I'm fine with that behavior overall, but it's a little hard to tell that that's what it's doing just from reading it. Should we have it print a separate debug message after it gives up or something like that?

Reply:
added a comment and tweaked the log line